flink-yarn-tests/src/test/java/org/apache/flink/yarn/YARNSessionCapacitySchedulerITCase.java

```diff
@@ -187,178 +188,146 @@ public void perJobYarnClusterOffHeap() throws IOException {
-			}
-			sleep(1000);
-		} while(!ok);
+		final List<ApplicationReport> apps = yarnClient.getApplications(EnumSet.of(YarnApplicationState.RUNNING));
+		assertEquals(1, apps.size()); // Only one running
+		return apps.get(0);
+	}
 
-		// send "stop" command to command line interface
-		runner.sendStop();
-		// wait for the thread to stop
-		try {
-			runner.join();
-		} catch (InterruptedException e) {
-			LOG.warn("Interrupted while stopping runner", e);
+	private static String normalizeTrackingUrl(final String trackingUrl) {
+		String url = trackingUrl;
+		if (!url.endsWith("/")) {
+			url += "/";
+		}
+		if (!url.startsWith("http://")) {
+			url = "http://" + url;
 		}
-		LOG.warn("stopped");
+		return url;
+	}
 
-		// ----------- Send output to logger
-		System.setOut(ORIGINAL_STDOUT);
-		System.setErr(ORIGINAL_STDERR);
-		oC = outContent.toString();
-		String eC = errContent.toString();
-		LOG.info("Sending stdout content through logger: \n\n{}\n\n", oC);
-		LOG.info("Sending stderr content through logger: \n\n{}\n\n", eC);
+	private void submitJob(final String jobFileName) throws IOException, InterruptedException {
+		Runner jobRunner = startWithArgs(new String[]{"run",
+				"--detached", getTestJarPath(jobFileName).getAbsolutePath()},
+			"Job has been submitted with JobID", RunTypes.CLI_FRONTEND);
+		jobRunner.join();
+	}
 
-		// ------ Check if everything happened correctly
-		Assert.assertTrue("Expect to see failed container",
-			eC.contains("New messages from the YARN cluster"));
+	private static void waitForTaskManagerRegistration(final String url, final Duration waitDuration) throws Exception {
+		waitUntilCondition(() -> getNumberOfTaskManagers(url) > 0, Deadline.fromNow(waitDuration));
+	}
 
-		Assert.assertTrue("Expect to see failed container",
-			eC.contains("Container killed by the ApplicationMaster"));
+	private static void assertNumberOfSlotsPerTask(final String url, final int slotsNumber) throws Exception {
+		try {
+			waitUntilCondition(() -> getNumberOfSlotsPerTaskManager(url) == slotsNumber, Deadline.fromNow(Duration.ofSeconds(30)));
+		} catch (final TimeoutException e) {
+			final int currentNumberOfSlots = getNumberOfSlotsPerTaskManager(url);
+			fail(String.format("Expected slots per TM to be %d, was: %d", slotsNumber, currentNumberOfSlots));
+		}
+	}
 
-		Assert.assertTrue("Expect to see new container started",
-			eC.contains("Launching TaskManager") && eC.contains("on host"));
+	private static int getNumberOfTaskManagers(final String url) throws Exception {
+		String response = TestBaseUtils.getFromHTTP(url + "taskmanagers/");
+		JsonNode parsedTMs = new ObjectMapper().readTree(response);
+		ArrayNode taskManagers = (ArrayNode) parsedTMs.get("taskmanagers");
+		return taskManagers == null ? 0 : taskManagers.size();
+	}
 
-		// cleanup auth for the subsequent tests.
-		remoteUgi.getTokenIdentifiers().remove(nmIdent);
+	private static int getNumberOfSlotsPerTaskManager(final String url) throws Exception {
```

Comment:
Here we could also use the `ClusterOverview` and simply divide the number of total slots by the number of TMs.